File: course/section_5_binary_search/code/08-search-in-rotated-sorted-array.py
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1
        while l <= r:
            m = (l + r) // 2
            mid = nums[m]
            if target == mid:
                return m
            if mid >= nums[l]:
                # left protion
                if mid <= target or target < nums[l]:
                    # have to sarch to the right
                    l = m + 1
                else:
                    r = m - 1
            else:
                # we are in the right portion
                if mid >= target or target > nums[r]:
                    # have to search the left
                    r = m - 1
                else:
                    # have to search the right
                    l = m + 1
        return -1

File: course/section_5_binary_search/code/08-search-in-rotated-sorted-array.php
<?php
class Solution {

    /**
     * @param Integer[] $nums
     * @param Integer $target
     * @return Integer
     */
    function search($nums, $target) {
        $left = 0  ;
        $right = count($nums) - 1 ;
        while($left <= $right){
            $m = intdiv($left + $right , 2) ;
            $mid = $nums[$m] ;
            if($mid == $target){
                return $m ;
            }
            if($mid >= $nums[0]){
                // left portion 
                if($mid < $target || $target < $nums[0]){
                    $left = $m + 1 ;
                } else {
                    $right = $m - 1 ;
                }
            } else {
                // right portion
                if($mid > $target || $target > $nums[count($nums) - 1]){
                    $right = $m - 1 ;
                } else {
                    $left = $m + 1 ;
                }
            }
        }
        return -1;
    }
}